File: python/evaluate_metrics.py
import numpy as np

def evaluate_metrics(y, y_pred, cm, metric, predisposition=None):
    if y is not None and y_pred is not None and cm is None:
        cm = computeCM(np.array(y), np.array(y_pred))
    elif (y is None or y_pred is None) and cm is not None:
        pass
    else:
        raise ValueError("Use either labels (y, y_pred) or a confusion matrix (cm), not both.")

    if predisposition is None:
        predisposition = 'average'

    if metric in ["AUROC", "ACscore", "acc"]:
        predisposition = 'positive'

    if predisposition == 'positive':
        perf = compute_metric(cm["tp"], cm["tn"], cm["fp"], cm["fn"], metric)

    elif predisposition == 'negative':
        flipped = flipcm(cm)
        perf = compute_metric(flipped["tp"], flipped["tn"],
                              flipped["fp"], flipped["fn"], metric)

    elif predisposition == 'average':
        a = compute_metric(cm["tp"], cm["tn"], cm["fp"], cm["fn"], metric)
        flipped = flipcm(cm)
        b = compute_metric(flipped["tp"], flipped["tn"],
                           flipped["fp"], flipped["fn"], metric)
        perf = np.nanmean([a, b])

    else:
        raise ValueError("predisposition must be 'positive', 'negative', or 'average'.")

    return perf


def flipcm(cm):
    return {
        "tp": cm["tn"],
        "tn": cm["tp"],
        "fp": cm["fn"],
        "fn": cm["fp"]
    }


def compute_metric(tp, tn, fp, fn, metric):
    if metric == "AUC":
        if (tp == 0 and fn == 0):
            sens = 1.0
        else:
            sens = tp / (tp + fn)

        if (fp == 0 and tn == 0):
            fpr = 0.0
        else:
            fpr = fp / (fp + tn)

        x = np.array([0, sens, 1], dtype=float)
        y = np.array([0, fpr, 1], dtype=float)
        result = np.trapz(x, y)

    elif metric == "AUPR":
        if (tp == 0 and fn == 0):
            sens = 1.0
        else:
            sens = tp / (tp + fn)

        if (tp == 0 and fp == 0):
            prec = 0.0
        else:
            prec = tp / (tp + fp)

        x = np.array([0, sens, 1], dtype=float)
        y = np.array([1, prec, 0], dtype=float)
        result = np.trapz(y, x)

    elif metric == "Fscore":
        if (tp == 0 and fn == 0):
            sens = 1.0
        else:
            sens = tp / (tp + fn)
        if (tp == 0 and fp == 0):
            prec = 0.0
        else:
            prec = tp / (tp + fp)

        if sens == 0.0 and prec == 0.0:
            result = 0.0
        else:
            result = 2.0 * prec * sens / (prec + sens)

    elif metric == "ACscore":
        if (tp == 0 and fn == 0):
            sens = 0.0
        else:
            sens = tp / (tp + fn)

        if (fp == 0 and tn == 0):
            spec = 0.0
        else:
            spec = tn / (fp + tn)

        if sens == 0.0 and spec == 0.0:
            result = 0.0
        else:
            result = 2.0 * sens * spec / (sens + spec)

    elif metric == "acc":
        total = tp + tn + fp + fn
        result = (tp + tn) / total if total > 0 else 0.0

    else:
        raise ValueError("Possible metrics: 'AUC', 'AUPR', 'Fscore', 'ACscore', 'acc'.")

    return float(result)


def computeCM(y, y_pred):
    y = np.array(y, dtype=int)
    y_pred = np.array(y_pred, dtype=int)

    tp = np.sum((y_pred == 1) & (y == 1))
    tn = np.sum((y_pred == 0) & (y == 0))
    fp = np.sum((y_pred == 1) & (y == 0))
    fn = np.sum((y_pred == 0) & (y == 1))

    return {
        "tp": int(tp),
        "tn": int(tn),
        "fp": int(fp),
        "fn": int(fn)
    }

File: matlab/src/utils/evaluate_metrics.m
function perf = evaluate_metrics(y, y_pred, tp, tn, fp, fn, metric, predisposition)

% Usage: 
% perf = evaluate_metrics(y, y_pred, [], metric, -) : evaluate from true
% labels y and predicted labels y_pred
% perf = evaluate_metrics([], [], cm, metric, -) : evaluate from confusion
% matrix cm

% if ~isempty(y) && ~isempty(y_pred) && isempty(cm)
%     cm = computeCM(y,y_pred);
% elseif isempty(y) && isempty(y_pred) && ~isempty(cm)
% 
% else
%     error("Use either labels or confusion matrix.")
% end

if nargin < 5
    predisposition = 'average';
end

if strcmp(metric, ["AUROC",'ACscore','acc'])
    predisposition = 'positive';
end

if strcmp(predisposition, 'positive')
    perf = compute_metric(tp, tn, fp, fn, metric);   % evaluation for positive class
elseif strcmp(predisposition, 'negative') % evaluation for negative class
    [tp, tn, fp, fn] = flipcm(tp, tn, fp, fn);
    perf = compute_metric(tp, tn, fp, fn, metric);
elseif strcmp(predisposition, 'average') % evaluation for both classes and take average
    a = compute_metric(tp, tn, fp, fn, metric);
    [tp, tn, fp, fn] = flipcm(tp, tn, fp, fn);
    b = compute_metric(tp, tn, fp, fn, metric);
    perf = mean([a,b], "omitnan"); 
end
end


function [tp, tn, fp, fn] = flipcm(tp, tn, fp, fn)  % flip both true labels and predicted labels.
tp0 = tp; tn0 = tn; fp0 = fp; fn0 = fn;
tp = tn0; tn = tp0; fp = fn0; fn = fp0;
end

function result = compute_metric(tp, tn, fp, fn, metric)
if strcmp(metric, 'AUC')
    if tp==0 && fn==0
        sens=1;
    else
        sens = tp/(tp+fn);
    end
    if fp==0 && tn==0
        fpr=0;
    else
        fpr = fp/(fp+tn);        
    end
    
    x = [0;sens;1];
    y = [0;fpr;1];
    
    result = trapz(y,x);
elseif strcmp(metric, 'AUPR')
    if tp==0 && fn==0
        sens=1;
    else
        sens = tp/(tp+fn);
    end
    if tp==0 && fp==0
        prec=0;
    else
        prec = tp/(tp+fp);
    end
    
    x = [0;sens;1];
    y = [1;prec;0];
    
    result = trapz(x,y);
elseif strcmp(metric, 'Fscore')
    sens = tp ./ (tp + fn);
    % sens(tp==0 & fn==0) = 1;

    prec = tp ./ (tp + fp);
    % prec(tp==0 & fp==0) = 0;

    result = (2 * prec .* sens) ./ (prec + sens);
    % result(sens==0 & prec==0) = 0;

    % result((tp==0 & fn==0) | (tp==0 & fp==0)) = 0;

    %%
    % if (tp==0 && fn==0) || (tp==0 && fp==0)
    %     result = 0;
    % else
    %     if tp==0 && fn==0
    %         sens=1;
    %     else
    %         sens = tp ./ (tp + fn);
    %     end
    %     if tp==0 && fp==0
    %         prec=0;
    %     else
    %         prec = tp ./ (tp + fp);
    %     end
    %     if sens==0 && prec==0
    %         result = 0;
    %     else
    %         result = (2 * prec .* sens) ./ (prec + sens);
    %     end
    % end
elseif strcmp(metric, 'ACscore')
    sens = tp ./ (tp + fn);
    % sens(tp==0 & fn==0) = 1;

    spc = tn ./ (fp + tn);
    % spc(fp==0 & tn==0) = 0;

    result = 2 * sens .* spc ./ (sens + spc);
    % result(sens == 0 & spc == 0) = 0;

    %%
    % if tp==0 && fn==0
    %     sens = 0;
    % else
    %     sens = tp/(tp+fn);
    % end
    % if fp==0 && tn==0
    %     spc = 0;
    % else
    %     spc = tn/(fp+tn);
    % end
    % if sens == 0 && spc == 0
    %     result = 0;
    % else
    %     result = 2*sens*spc/(sens+spc);
    % end
elseif strcmp(metric, 'acc')
    result = (tp + tn) / (tp + tn + fp + fn);
else
    error('Possible metrics: ''AUROC'', ''AUPR'', ''Fsore'', ''ACscore'', ''acc''.');
end
end


function cm = computeCM(y,y_pred)
cm = allocateCM();
cm.tp = sum(y_pred == 1 & y == 1);
cm.tn = sum(y_pred == 0 & y == 0);
cm.fp = sum(y_pred == 1 & y == 0);
cm.fn = sum(y_pred == 0 & y == 1);
end

function cm = allocateCM()
cm.tp = zeros(1,"uint64");
cm.tn = zeros(1,"uint64");
cm.fp = zeros(1,"uint64");
cm.fn = zeros(1,"uint64");
end